src/util/environment.ts

```diff
@@ -0,0 +1,332 @@
+    browser = "Microsoft Internet Explorer";
+    browserVersion = navigator.userAgent.substring(verOffset + 5);
+  }
+
+  //IE 11 no longer identifies itself as MS IE, so trap it
+  //http://stackoverflow.com/questions/17907445/how-to-detect-ie11
+  else if (
+    browser == "Netscape" &&
+    navigator.userAgent.indexOf("Trident/") != -1
+  ) {
+    browser = "Microsoft Internet Explorer";
+    browserVersion = navigator.userAgent.substring(verOffset + 5);
+    if ((verOffset = navigator.userAgent.indexOf("rv:")) != -1) {
+      browserVersion = navigator.userAgent.substring(verOffset + 3);
+    }
+  }
+
+  // Chrome
+  else if ((verOffset = navigator.userAgent.indexOf("Chrome")) != -1) {
+    browser = "Chrome";
+    browserVersion = navigator.userAgent.substring(verOffset + 7);
+  }
+  // Safari
+  else if ((verOffset = navigator.userAgent.indexOf("Safari")) != -1) {
+    browser = "Safari";
+    browserVersion = navigator.userAgent.substring(verOffset + 7);
+    if ((verOffset = navigator.userAgent.indexOf("Version")) != -1) {
+      browserVersion = navigator.userAgent.substring(verOffset + 8);
+    }
+
+    // Chrome on iPad identifies itself as Safari. Actual results do not match what Google claims
+    //  at: https://developers.google.com/chrome/mobile/docs/user-agent?hl=ja
+    //  No mention of chrome in the user agent string. However it does mention CriOS, which presumably
+    //  can be keyed on to detect it.
+    if (navigator.userAgent.indexOf("CriOS") != -1) {
+      //Chrome on iPad spoofing Safari...correct it.
+      browser = "Chrome";
+      //Don't believe there is a way to grab the accurate version number, so leaving that for now.
+    }
+  }
+  // Firefox
+  else if ((verOffset = navigator.userAgent.indexOf("Firefox")) != -1) {
+    browser = "Firefox";
+    browserVersion = navigator.userAgent.substring(verOffset + 8);
+  }
+  // Other browsers
+  else if (
+    (nameOffset = navigator.userAgent.lastIndexOf(" ") + 1) <
+    (verOffset = navigator.userAgent.lastIndexOf("/"))
+  ) {
+    browser = navigator.userAgent.substring(nameOffset, verOffset);
+    browserVersion = navigator.userAgent.substring(verOffset + 1);
+    if (browser.toLowerCase() == browser.toUpperCase()) {
+      browser = navigator.appName;
+    }
+  }
+  // trim the browser version string
+  if ((ix = browserVersion.indexOf(";")) != -1)
+    browserVersion = browserVersion.substring(0, ix);
+  if ((ix = browserVersion.indexOf(" ")) != -1)
+    browserVersion = browserVersion.substring(0, ix);
+  if ((ix = browserVersion.indexOf(")")) != -1)
+    browserVersion = browserVersion.substring(0, ix);
+
+  return [browser, browserVersion].join("-");
+};
+
+/**
+ * Get OS details for the browser
+ */
+const getBrowserOsDetails = (
+  navigator: Navigator | WorkerNavigator
+): string => {
+  let os = "unknown";
+  const clientStrings = [
+    { s: "Windows 10", r: /(Windows 10.0|Windows NT 10.0)/ },
```

Comment:
windows 11 is the latest windoes and is not present in this list

Reply:
good catch! I am looking into this and my understanding is now that it is not possible to distinguish between Windows 10 and 11 through the user agent string. Chromium is freezing user agents strings in favor of newerm methods, which I think means the [User-Agent Client Hints API](https://developer.mozilla.org/en-US/docs/Web/API/User-Agent_Client_Hints_API). Unfortunately though, it's still in an experimental state and at least firefox doesn't support said methods yet.

https://developers.whatismybrowser.com/learn/browser-detection/user-agents/detect-windows-11-user-agent

I suggest we proceed as is, and consider following up with something like:

1. check if the user-agent hints API is available and use that
2. fall back on current impl